enhancements/network/baremetal-ipi-network-configuration.md

```diff
@@ -0,0 +1,187 @@
+due to the chicken/egg problem with accessing the MCS.
+
+#### Story 2
+
+As an on-premise IPI user, I wish to use static IPs for my controlplane network,
+for reasons of network ownership or concerns over reliability I can't use DHCP
+and therefore need to provide a static configuration for my primary network.
+
+There is no way to provide [MachineSpecific Configuration in OpenShift](https://github.com/openshift/machine-config-operator/issues/1720) so I am
+forced to use the UPI flow which is less automated and more prone to errors.
+
+### Risks and Mitigations
+
+In some existing workflows, kubernetes-nmstate is used to do network configuration on day-2. Using a different interface for day-1 introduces the potential for mismatches and configuration errors when making day-2 changes. It should be possible to mitigate this by generating the NetworkManager keyfiles using the nmstatecli, which will allow the same nmstate config to be used directly in the day-2 configuration.
+
+## Design Details
+
+In the IPI flow day-1 network configuration is required in 2 different cases:
+
+* Deployment of the controlplane hosts via terraform, using input provided to openshift-install
+* Deployment of compute/worker hosts (during initial deployment and scale-out), via Machine API providers for each platform
+
+In the sections below we will describe the user-facing API that contains network configuration, and the proposed integration for each of these cases.
+
+### User-facing API
+
+RHCOS already provides a mechanism to specify NetworkManager keyfiles during deployment of a new node. We need to expose that functionality during the IPI install process. There are a couple of options on how to do that:
+
+* A new section in install-config.
+* A secret that contains base64-encoded content for the keyfiles.
+
+These are not mutually exclusive. If we implement the install-config option, we will still need to persist the configuration in a secret so it can be used for day-2. In fact, the initial implementation should likely be based on secrets, and then install-config integration can be added on top of that to improve the user experience. This way the install-config work won't block implementation of the feature.
+
+The data provided by the user will need to have the following structure:
+```
+[hostname]:
+  [interface name]:
+    [contents of NetworkManager key file for interface]
+  [second interface name]:
+    [contents of NetworkManager key file for second interface]
+  etc...
+```
+
+For example:
+```
+master-0:
+  eth0:
+    [connection]
+    id=eth0
+    uuid=18e0cec7-041c-4fb3-957c-a60c80dd9b85
+    type=ethernet
+    etc...
+master-1:
+  eth0:
+    [connection]
+    id=eth0
+    etc...
+```
+
+#### API Open Questions
+
+* How do we map an individual machine's keyfiles to the machine? Hostname was previously suggested because it was already an available label in NNCP, but if we're not using NNCP then there might be a better way.
+
```

Comment:
I think the current plan is to initially enable this interface only for the `baremetal` platform, so there may be concerns about adding a new top-level install-config section.  

In that case, we could instead add the config to the platform-specific `hosts` list (which already contains other host-specific data needed to automate deployment), and this also solves the mapping to a host (since we're already consuming this per-host data in the `baremetal` flow, e.g BMC credentials)

Reply:
I agree with Steve. This should belong in the hosts section of the baremetal platform. e.g.
```
platform:
  baremetal:
    hosts:
      - name: openshift-master-0
        networkData:
          eth0:
            [connection]
            id=eth0
            uuid=18e0cec7-041c-4fb3-957c-a60c80dd9b85
            type=ethernet
etc...     
```

Reply:
Okay, this sounds reasonable. I'd be more concerned about using a baremetal-specific structure if we had the common nmstate API, but since this implementation isn't going to be useful to other groups like OVNK anyway I'm good with it.

Reply:
I guess that raises the question (already mentioned in the alternatives) of whether we want this to be nmstate format or just inlined keyfiles?  I guess we need the interface to enable adding nmstate support even if we start with keyfiles.

Also for those familiar with OpenStack `networkData` implies a different format - maybe we should call this something else e.g `networkConfig` or `networkFiles`  (then adding e.g `nmstateConfig` later would be simple)?

Reply:
If we use nmstate yaml for install-config then we need nmstate available on the provisioning host so we can convert it to keyfiles. I'm fine with doing that (it's how we'd have to process NNCP records too), but it does add a dependency. I'm not sure how controversial that will be. It's also a bit more complex implementation-wise than straight keyfiles.

One option would be to replace the interface name with the filename, i.e. eth0 -> eth0.nmconnection. That's going to have to happen at some point anyway, and then we could potentially add support for eth0.yaml, which would indicate nmstate data. But maybe that's a little too magic?

+1 to not overloading networkData.

Reply:
That's a good point - given that it's python based I think it's highly unlikely we'll get the nmstate CLI into RHCOS - we'd need to run it inside a container.

As @dankenigsberg mentions below one advantage of using nmstate would be it may be easier to handle changes in NM configuration syntax after upgrades, but until #201 is resolved we don't actually lifecycle bootimages on upgrade (presumably this will be needed before any el8->el9 based RHCOS upgrade though)

On balance I'm leaning towards starting with keyfiles, but then adding nmstate support in a future release - we can bikeshed the interface later but yes the connection suffix does sound a little magical :)  I'd maybe prefer a new section to support the alternative syntax

Reply:
Starting with keyfiles makes a lot of sense. On some level we'll need keyfiles for day 1 since that's what we'll pass in to the images. Layering another interface over that should be much simpler once we have keyfile support.

Reply:
+1 for filenames to replace the interface name in the API.

Reply:
We can probably simplify the config a bit and also use the suggestion to include file names like so:

```yaml
platform:
  baremetal:
    hosts:
      - name: openshift-master-0
        networkConfig:
        - eth0.nmconnection: |
            [connection]
            id=eth0
            uuid=18e0cec7-041c-4fb3-957c-a60c80dd9b85
            type=ethernet
            etc...
```

Reply:
I also wanted to clarify which format we're all leaning towards: NM Keyfiles or nmstate config. Either way, I hope I hope we're just consuming a blob of network config data. I was just looking at the nmstate schema and its pretty extensive. I don't think we should include it in the install-config. At least in this proposal. 
https://github.com/nmstate/nmstate/blob/base/libnmstate/schema.py

Reply:
Do we actually forsee the installer trying to parse/validate this data?  That seems hard.

Ultimately this data is just passed through into an Ignition config provided to the node, right?

I really see this as leading towards a general ability to inject ignition (rendered indirectly perhaps via [butane](https://github.com/coreos/butane/)) or other tools to specific hosts.

xref https://github.com/openshift/machine-config-operator/issues/1720

Regarding nmstate specifically; I think this actually has the same issue as e.g. offering `butane` configs.
It may be simplest to have `openshift-install` only accept final rendered ignition configs to start.  Then an administrator who wants to generate NM configs via nmstate (or use `butane` or another custom tool) needs to run that outside of `openshift-install`.

Reply:
> Do we actually forsee the installer trying to parse/validate this data? That seems hard.

No, I don't think the installer would validate anything beyond the fact that it's valid yaml.

> Ultimately this data is just passed through into an Ignition config provided to the node, right?

Yes.

> I really see this as leading towards a general ability to inject ignition (rendered indirectly perhaps via [butane](https://github.com/coreos/butane/)) or other tools to specific hosts.

Not sure I agree here. The network data is only one part of the ignition we give the host. The rest involves pulling the ironic agent container and running it as a service. There's no way we want users to even have to know that exists.